File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/state/UseSmileIDSampleIdDetails.kt
package com.usesmileid.sampleapps.ui.state

import androidx.compose.runtime.Immutable

/** The ID-details form. ID type stays unselectable until a country is chosen, which the form is built around. */
@Immutable
data class UseSmileIDSampleIdDetails(
    val country: UseSmileIDSampleCountry? = null,
    val idType: UseSmileIDSampleIdType? = null,
    val idNumber: String = "",
) {
    val isComplete: Boolean get() = country != null && idType != null && idNumber.isNotBlank()
}

/** A country the sandbox supports, with the flag the picker leads each row with. */
enum class UseSmileIDSampleCountry(val code: String, val label: String, val flag: String) {
    Nigeria("NG", "Nigeria", "🇳🇬"),
    Kenya("KE", "Kenya", "🇰🇪"),
    Ghana("GH", "Ghana", "🇬🇭"),
    SouthAfrica("ZA", "South Africa", "🇿🇦"),
    Uganda("UG", "Uganda", "🇺🇬"),
    Tanzania("TZ", "Tanzania", "🇹🇿"),
    Rwanda("RW", "Rwanda", "🇷🇼"),
}

/** ID types, filtered by country, which is why the trigger is disabled until one is chosen. */
enum class UseSmileIDSampleIdType(val id: String, val label: String, val countries: Set<UseSmileIDSampleCountry>) {
    NationalId("NATIONAL_ID", "National ID", UseSmileIDSampleCountry.entries.toSet()),
    Passport("PASSPORT", "Passport", UseSmileIDSampleCountry.entries.toSet()),
    DriversLicense("DRIVERS_LICENSE", "Driver's licence", setOf(UseSmileIDSampleCountry.Nigeria, UseSmileIDSampleCountry.Kenya, UseSmileIDSampleCountry.SouthAfrica)),
    VoterId("VOTER_ID", "Voter ID", setOf(UseSmileIDSampleCountry.Nigeria, UseSmileIDSampleCountry.Ghana)),
    ;

    companion object {
        fun of(country: UseSmileIDSampleCountry?) = entries.filter { country != null && country in it.countries }
    }
}

File: ios/SampleUI/Sources/SampleUI/State/UseSmileIDSampleIdDetails.swift
/// The ID-details form. ID type stays unselectable until a country is chosen.
public struct UseSmileIDSampleIdDetails: Equatable, Sendable {
  public var country: UseSmileIDSampleCountry?
  public var idType: UseSmileIDSampleIdType?
  public var idNumber: String

  public init(
    country: UseSmileIDSampleCountry? = nil,
    idType: UseSmileIDSampleIdType? = nil,
    idNumber: String = ""
  ) {
    self.country = country
    self.idType = idType
    self.idNumber = idNumber
  }

  public var isComplete: Bool {
    country != nil && idType != nil && !idNumber.isBlank
  }
}

/// A country the sandbox supports, with the flag the picker leads each row with.
public enum UseSmileIDSampleCountry: String, CaseIterable, Sendable {
  case nigeria = "NG"
  case kenya = "KE"
  case ghana = "GH"
  case southAfrica = "ZA"
  case uganda = "UG"
  case tanzania = "TZ"
  case rwanda = "RW"

  public var code: String {
    rawValue
  }

  public var label: String {
    switch self {
    case .nigeria: "Nigeria"
    case .kenya: "Kenya"
    case .ghana: "Ghana"
    case .southAfrica: "South Africa"
    case .uganda: "Uganda"
    case .tanzania: "Tanzania"
    case .rwanda: "Rwanda"
    }
  }

  public var flag: String {
    switch self {
    case .nigeria: "\u{1F1F3}\u{1F1EC}"
    case .kenya: "\u{1F1F0}\u{1F1EA}"
    case .ghana: "\u{1F1EC}\u{1F1ED}"
    case .southAfrica: "\u{1F1FF}\u{1F1E6}"
    case .uganda: "\u{1F1FA}\u{1F1EC}"
    case .tanzania: "\u{1F1F9}\u{1F1FF}"
    case .rwanda: "\u{1F1F7}\u{1F1FC}"
    }
  }
}

/// ID types, filtered by country, which is why the trigger is disabled until one is chosen.
public enum UseSmileIDSampleIdType: String, CaseIterable, Sendable {
  case nationalId = "NATIONAL_ID"
  case passport = "PASSPORT"
  case driversLicense = "DRIVERS_LICENSE"
  case voterId = "VOTER_ID"

  public var id: String {
    rawValue
  }

  public var label: String {
    switch self {
    case .nationalId: "National ID"
    case .passport: "Passport"
    case .driversLicense: "Driver's licence"
    case .voterId: "Voter ID"
    }
  }

  public var countries: Set<UseSmileIDSampleCountry> {
    switch self {
    case .nationalId, .passport: Set(UseSmileIDSampleCountry.allCases)
    case .driversLicense: [.nigeria, .kenya, .southAfrica]
    case .voterId: [.nigeria, .ghana]
    }
  }

  public static func of(_ country: UseSmileIDSampleCountry?) -> [UseSmileIDSampleIdType] {
    guard let country else { return [] }
    return allCases.filter { $0.countries.contains(country) }
  }

  public static func of(_ country: UseSmileIDSampleCountry?, matching query: String) -> [UseSmileIDSampleIdType] {
    of(country).filter { $0.label.matches(query) }
  }
}

public extension UseSmileIDSampleCountry {
  static func matching(_ query: String) -> [UseSmileIDSampleCountry] {
    allCases.filter { $0.label.matches(query) }
  }
}

extension String {
  /// An empty query matches everything. `localizedCaseInsensitiveContains("")` answers false, where
  /// the Compose twin's `contains("")` answers true — so a literal port opens both pickers empty.
  func matches(_ query: String) -> Bool {
    query.isBlank || localizedCaseInsensitiveContains(query)
  }
}
